### PR TITLE
Improve #inspect format for Time types

### PR DESCRIPTION
**Pull request by straight-shoota**
I'm doing some housekeeping to normalize and improve the inspect (and to_s) output of time and location related types.

`Time#inspect` now includes nanoseconds <del>unless it is `0`</del>. `#to_s` still omits them to avoid cluttering the output. When calling `#inspect` you typically expect a detailed output that allows to distinguish the difference between two instances. While nanoseconds are usually not relevant, it should still show up in case it is significant.

```crystal
Time.now.inspect # => 2018-03-09 12:52:36.506736000 +01:00 Europe/Berlin
Time.now.to_s    # => 2018-03-09 12:52:36 +01:00 Europe/Berlin
```

`Time::Zone` now includes a `#format` method which is used to print the offset as `+HH:mm:ss`. This format is added to the `#inspect` output and also used by `Time::Formatter#time_zone`.

### CODE REVIEW

**Comment by RX14** [comment]
I'd expect inspect to include nanoseconds even if nanoseconds was 0.

**Comment by straight-shoota** [comment]
IMO it just adds unnecessary clutter. And `nanoseconds == 0` is pretty common because it is usually not included when serializing time data.

But I see the point, that it adds uncertainty when a time instance is printed, because it can mean nanoseconds is just omitted (e.g. when calling `to_s`) or actually `0`.

What do you think about printing only one decimal place (`.0`) if it is `0`?

**Comment by RX14** [comment]
@straight-shoota yes that's fine

File: spec/std/time/time_spec.cr [hunk 1]
```diff
@@ -283,56 +283,65 @@ describe Time do
...
+      Time.utc(2014, 10, 30, 21, 1, 13).to_s.should eq("2014-10-30 21:01:13 UTC")
+      Time.utc(2014, 10, 30, 21, 18, 1).to_s.should eq("2014-10-30 21:18:01 UTC")
     end
 
-    it "prints UTC" do
-      t = Time.utc 2014, 10, 30, 21, 18, 13
-      t.to_s.should eq("2014-10-30 21:18:13 UTC")
+    it "omits nanoseconds" do
+      Time.utc(2014, 10, 30, 21, 18, 13).to_s.should eq("2014-10-30 21:18:13 UTC")
+      Time.utc(2014, 10, 30, 21, 18, 13, nanosecond: 12345).to_s.should eq("2014-10-30 21:18:13 UTC")
```
Comment (ysbaddaden) [inline spec/std/time/time_spec.cr]: Why should `Time#to_s` omit nanoseconds when it's not zero?

**Comment by straight-shoota** [comment]
ISO 8601 or RFC 3339 only support fixed offsets. `Time` instances cannot be entirely expressed in those formats, because they are essentially an ISO 8601 date with an attached location which provides zone transition rules.

Two instances of `Time` having different locations but sharing the same current offset should be distinguishable. They behave differently for example when adding a time span. Thus `location` is significant for inspecting a `Time` instance.

If time zone rules don't matter, you should just use `UTC` and/or fixed offset locations if necessary. Such instances can be expressed as ISO 8601 or RFC 3339 and the location is omitted in `#inspect`.

**Comment by ysbaddaden** [comment]
My review got lost, but in essence:

- `#inspect` printing a dangling `.0` is useless noise;
- `#to_s` has no reason to skip nanoseconds (unless zero);
- `#to_s` should have a `location` argument that defaults to `Location::Local`, allowing useful constructs such as `Time.with(user.location) { puts comment.created_at.to_s }`; bonus: we don't have to print a location.

Actually, I believe all Time instances should be shifted to localtime —I'm probably wrong, but I never got a use-case where keeping the original location was essential or even useful. The fact that international standards (ISO, RFC) don't specify locations is a huge hint to me that what's important is the offset (structural deviation from UTC), not the location (context specific representational concept).

**Comment by straight-shoota** [comment]
I now concur with @RX14 that nanoseconds should better be explicit. It's obviously very unlikely that this gets confused with a custom display of a non zero-nanosecond time (for example `time.to_s("%F %T %Z #{time.location}")), but I think we should better be safe than sorry. IMHO the additional noise is acceptable.

I'm not sure about `to_s`... could be with or without nanoseconds, I don't have a clear position if it should.

I like the idea of specifying a location for formatting, but I would just add it as an argument to `#to_s`:
`comment.created_at.to_s(user.location)`. This would format the time in the given location without printing the actual location. It could probably even omit the offset because it is clearly a localized format.
It's essentially the same as `comment.created_at.in(user.location).to_s` but with a modified default format.

What do you mean with "all Time instances should be shifted to localtime"? `Time#location` should always be `Location.local`??

The location is usually not that important, that's true. But it depends on what you're doing with your `Time` instances. If it's just recording timed events and printing that in a localized format, then yes, you don't need the `Time` instance to have any location. It should just be UTC and it is only localized when printed.
But if you're doing calendar operations such as adding time spans, it is essential to have the `Time` instance in a specific location.

**Comment by straight-shoota** [comment]
After thinking about this, I'm pretty sure `to_s` should *not* print nanoseconds by default. Many `Time` instances will have a non-zero nanoseconds value because they were created as `Time.now` but for the great majority of use cases sub-second precision is completely irrelevant. Let's keep the default format short and precise for ~90% of use cases. It is always possible to specify a custom format for `to_s`.

**Comment by RX14** [comment]
nah, we should keep every detail including nanoseconds and zone on `inspect` and have `to_s` just be `date time +-offset` without ns.

**Comment by straight-shoota** [comment]
@RX14's suggestion seems reasonable, `to_s` gives the typically needed information and `inspect` shows everything.

**Comment by RX14** [comment]
needs rebase for CI fix

**Comment by straight-shoota** [comment]
Rebased

File: src/time.cr [hunk 2]
```diff
@@ -471,21 +471,51 @@ struct Time
...
-  def inspect(io : IO)
-    case
-    when utc?
-      to_s "%F %T UTC", io
-    else
-      if offset % 60 == 0
-        to_s "%F %T %:z", io
+  # Prints this `Time` to *io*.
+  #
+  # The local date-time is formatted as ISO 8601 date string `YYYY-MM-DD HH:mm:ss.nnnnnnnnn +ZZ:ZZ:ZZ`.
```
Comment (RX14) [inline src/time.cr]: that string isn't iso 8601 and the example doesn't include the zone name.

File: src/time.cr [hunk 3]
```diff
@@ -471,21 +471,51 @@ struct Time
...
+      zone.format(io)
+      io << ' ' << location.name unless location.fixed?
+    end
+
     io
   end
 
+  # Prints this `Time` to *io*.
+  #
+  # The local date-time is formatted as ISO 8601 date string `YYYY-MM-DD HH:mm:ss +ZZ:ZZ:ZZ`.
```
Comment (RX14) [inline src/time.cr]: also not iso 8601.

File: src/time.cr [hunk 4]
```diff
@@ -471,21 +471,51 @@ struct Time
...
+  # Nanoseconds are always omitted. Offset seconds are omitted if `0`.
+  #
+  # The name of the location is appended unless it is a fixed zone offset.
```
Comment (RX14) [inline src/time.cr]: It shouldn't be.

File: src/time.cr (hunk 2 above)
Comment (straight-shoota) [reply src/time.cr]: Why isn't it ISO 8601?

Comment (straight-shoota) [reply src/time.cr]: Zone name is mentioned in the next paragraph.

Comment (RX14) [reply src/time.cr]: no T, and there's a space between the nanoseconds and the +

Comment (straight-shoota) [reply src/time.cr]: `T` designator can be omitted, but the additional white space is true. The question is: should it be removed?

**Comment by straight-shoota** [comment]
Merge and milestone?

**Pull request merged**